Refactor benchmark tools for statistical significance

AndreasMadsen opened the pull request:
##### Checklist

- [x] tests and code linting passes
- [ ] a test and/or benchmark is included
- [x] documentation is changed or added
- [x] the commit message follows commit guidelines

##### Affected core subsystem(s)
benchmark

##### Description of change

I have been rather confused about the benchmark suite and I don't think it is as user friendly as the rest of nodecore. This PR attempt to remove most of the confusion I was facing when I started using it. Primarily it:

* removes unused/undocumented files
* allows partially setting the benchmarks variables using process arguments.
* refactor `compare.js` such comparing node versions and getting statistical significance is easy.
* refactor the `plot.R` tool (now called `scatter`) to show a scatter plot with confidence bars.
* refactor cli tools such the cli API is more homogeneous.
* documents all the tools.
* removes the implicit `process.exit(0)` after `bench.end()`.
* uses `process.send` to avoid most parsing (the benchmark cli arguments haven't changed).

The specifics are documented in the commit messages. Please also see the the [new README](https://github.com/AndreasMadsen/node/blob/benchmark-refactor/benchmark/README.md) as quite a lot have changed (be sure the to check my spelling!).

_Note that some benchmark takes a very long time to complete, e.g. `timers/timers.js	type=depth thousands=500` takes 11.25 min. Thus running it 30 times for statistical significance is unreasonable. I suspect the only reason why it is set to so many iterations is to get a small variance, but with the the new `compare` tool the variance can be estimated instead of being reduced. Thus we can reduce the number of iterations and still get the information we need. But I suggest we do that in another pull request, as is very different discussion._

**Motivation (long story)**: I wanted to benchmark the effect of some `async_wrap` changes. I went to the `benchmark/` directory and read the [README](https://github.com/nodejs/node/blob/master/benchmark/README.md). However I quickly discovered that it was primarily about running benchmarks a single time and how to write benchmarks. And most importantly it didn't explain how to compare two node versions. _This is now documented in the [new README](https://github.com/AndreasMadsen/node/blob/benchmark-refactor/benchmark/README.md)._

I then had to search for the tools myself and discovered the large amount of benchmarks files which where not put into categorized directories. I assumed they where somehow extra significant, but in reality they just appear to be unused. _These files are now removed._

After discovering the compare tool, which has the cli API
```
node benchmark/compare.js
            <node-binary1> <node-binary2> +
            [--html] [--red|-r] [--green|-g] +
            [-- <type> [testFilter]]
```
I was confused about what the `--red`, `--green` was and how the `node-binary1` and `node-binary2` compared, should I write `./node-old ./node-new` or `./node-new ./node-old` if I wanted a positive improvement factor to signify an improvement? _The new compare API is_:
```
usage: ./node benchmark/compare.js <type> ...
  --new    ./new-node-binary  new node binary (required)
  --old    ./old-node-binary  old node binary (required)
  --runs   30                 number of samples
  --filter pattern            string to filter benchmark scripts
  --var    variable=value     set benchmark variable (can be repeated)
```

After understanding `common.js` this it was still unclear if the performance was statistically significant different. I tried running the benchmark 5 times and got that 4/5 was an improvement, I was expecting it to have the same performance or be slower. (spoiler: it wasn't significant). _The compare.js script now runs the benchmarks many times (30 by default) and there is an R script to analyse the csv results._

At this point I wanted to do a rewrite of the benchmark tools (not the benchmarks themself) and changed a few other things in the process as well. - I'm a mathematician so I care a lot about statistical significance :)

AndreasMadsen commented:
I'm not sure who to cc for this one.
/cc @Trott as you appear to have made some resent benchmark changes.

Trott commented:
@nodejs/benchmarking

Trott commented:
In theory, this sounds fantastic to me! In practice, there's so much about benchmarking that I'm ignorant about, I have to defer to others.

jasnell commented:
Very nice. @mscdex @bnoordhuis ... any thoughts on this?

mscdex commented on benchmark/misc/string-creation.js:
```diff
@@ -1,17 +0,0 @@
-  var s;
-  for (var i = 0; i < n; i++) {
-    s = '01234567890';
-    s[1] = 'a';
```
Perhaps this line was added to prevent v8 from optimizing the for-loop away or something (since `s` wouldn't have been referenced)?

AndreasMadsen replied:
Perhaps. With `use strict` it is definitely broken. Looking at the original commit ( https://github.com/nodejs/node/commit/12a169e8ff4036e160dbdc605eb9f15eae933b47 - 6 years ago) it seams like it was just a misunderstanding of how strings works. The commit appears to compare strings and buffers, which is not comparable in this case as strings are immutable.

mscdex commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
+  // Construct confiuration string, " A=a, B=b, ..."
+  let conf = '';
+  for (const key of Object.keys(data.conf)) {
+    conf += ' ' + key + '=' + data.conf[key];
```
`data.conf[key]` may need to be `JSON.stringify()`ed for strings with control characters (including newlines, etc.), otherwise you'll end up with messed up output. I had to change this with the existing benchmark runner when benchmarking some http header character validation functions.

AndreasMadsen replied:
I'm happy to do that. But I don't want to change the format in case it is used by some 3th party benchmark monitoring setup.

mscdex commented on benchmark/run.js:
```diff
@@ -1,63 +1,43 @@
+    // Construct configuration string, " A=a, B=b, ..."
+    let conf = '';
+    for (const key of Object.keys(data.conf)) {
+      conf += ' ' + key + '=' + data.conf[key];
```
Same thing here about `JSON.stringify()`ing `data.conf[key]`.

mscdex commented on benchmark/run.js:
```diff
@@ -1,63 +1,43 @@
+const fork = require('child_process').fork;
+const CLI = require('./_cli.js');
+
+const cli = CLI(`usage: ./node benchmark/run.js <type> ...
```
There should probably be some explanation (in the help text) about what `<type>` should be exactly...

mscdex commented:
Just briefly looking over it, it mostly seems to look ok except for a few nits.

I did spot a typo in the `benchmark: add script for creating scatter plot` commit message body.

mscdex commented on benchmark/compare.js:
```diff
@@ -1,181 +1,73 @@
+//
+// Parse arguments
+//
+const cli = CLI(`usage: ./node benchmark/compare.js <type> ...
```
There should probably be some explanation (in the help text) about what <type> should be exactly...

AndreasMadsen replied:
are you referring to `<type>`?

mscdex replied:
Yes, markdown cut that part out.

trevnorris replied:
iirc there was another one of these in another commit. just to watch out for it.

AndreasMadsen replied:
They should all be fixed. Unless you are talking about the R scripts, but they only take `--` arguments.

addaleax replied:
btw, the first few times I tried running the new I found it *very* confusing that the `type` argument needed to appear before the arguments starting with `--` (i.e. `compare.js --new bla --old blah http` did not work). I almost never use CLIs with that argument order, and just showing this usage text wasn’t exactly helpful, either.

You don’t need to change the behaviour, but maybe add a note here about that and for the other scripts where it applies?

AndreasMadsen replied:
Can you elaborate on that note, I think it is very specific, this is the message you get now.

```
usage: ./node benchmark/compare.js <type> ...
  Run each benchmark in the <type> directory many times using two diffrent
  node versions. More than one <type> directory can be specified. The output is
  formatted as csv, which can be processed using for example 'compare.R'.

  --new    ./new-node-binary  new node binary (required)
  --old    ./old-node-binary  old node binary (required)
  --runs   30                 number of samples
  --filter pattern            string to filter benchmark scripts
  --set    variable=value     set benchmark variable (can be repeated)
```

I choose this order, because it could be implemented using less code.

I will try and change the argument order, this appears to cause a lot of confusion for many people, but I would love to understand why.

addaleax replied:
> I will try and change the argument order, this appears to cause a lot of confusion for many people, but I would love to understand why.

If I had to guess, I’d say it’s because that’s the order usually suggested in man pages and `--help` texts, and maybe because the positional arguments are the ones one is most likely to spend more time editing before hitting enter… idk, maybe there’s more to it.

AndreasMadsen replied:
Oh I understand the order is confusing (it is fixed now). But this is the third comment I got about a missing note, but unless I'm misunderstanding the comment, there is note just one line below.

AndreasMadsen commented:
@mscdex thanks. Updated as suggested.

AndreasMadsen commented:
ping

mscdex commented on benchmark/README.md:
````diff
@@ -1,147 +1,276 @@
 ```
 
-## How to write a benchmark test
+After generating the csv, a comparens table can be created using the `scatter.R`
````
s/comparens/comparison ?

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
+
+## Creating a benchmark
+
+All benchmarks uses the `require('../common.js')` module. This contains the
```
s/uses/use/

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
-var bench = common.createBenchmark(main, {
-  type: ['fast', 'slow'], // Two types of buffer
-  n: [512] // Number of times (each unit is 1024) to call the slice API
+The first argument `main` is the benchmark function, the second arguments
```
s/arguments/argument/

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
+The first argument `main` is the benchmark function, the second arguments
+specifies the benchmark parameters. `createBenchmark` will run all possible
+combinations of these parameters, unless specified otherwise. Note that the
+configuration values can only be strings and numbers.
```
s/and/or/

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
+available through your preferred package manager. If not `wrk` can be built
+[from source][wrk] via `make`.
+
+The R scripts uses `ggplot2` and `plyr`, you can install them using from the
```
s/uses/use/

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
+conclusions about the performance.
 
-### Run all tests of a given type
+individual benchmarks can be executed by simply executing the benchmark script
```
s/individual/Individual/

AndreasMadsen replied:
search and replace part is the same? Looked it up in oxford, it appears `individual` is the correct spelling.

thefourtheye replied:
First letter should be a capital letter.

mscdex replied:
Yes, sorry, that is what I meant. First letter should be capitalized.

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
-the test function with each of the combined arguments in spawned processes. For
-example, buffers/buffer-read.js has the following configuration:
+Each line represents a single benchmark with parameters specified as
+`${variable}=${value}`. Each configuration combination is executed in separate
```
s/in separate processes/in a separate process/

AndreasMadsen replied:
Thanks. I'm really curious why, do you know the name of the rule?

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
-The last number is the rate of operations. Higher is better.
+### Comparing node versions
+
+To compare the effect of a new node versions use the `compare.js` tool. This
```
s/versions/version/

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
 
-For example, buffer-slice.js:
+First build two versions of node, one from the master branch (here called
+`./node-mater`) and another with the pull request applied (here called
```
s/mater/master/

mscdex commented on benchmark/compare.R:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  r = list(
+    improvment = improvment,
```
s/improvment/improvement/ for all instances in this script and in the new benchmark document changes

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
 
-This example will run only the first type of url test, with one iteration.
-(Note: benchmarks require __many__ iterations to be statistically accurate.)
+The `compare.R` tool can also produces a box plot by using the `--plot filename`
```
s/produces/produce/

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
-(Note: benchmarks require __many__ iterations to be statistically accurate.)
+The `compare.R` tool can also produces a box plot by using the `--plot filename`
+option. In this case there are 48 different benchmark combinations, thus you
+may want to filter the csv file. This can be during benchmarking using for
```
s/be during/be done while/
This sentence might be totally reworded though, it sounds a bit awkward as-is. Perhaps something like:

```
This can be done while benchmarking using the `--var` parameter (e.g. `--var encoding=ascii`) or by filtering output afterwards using tools such as `sed` or `grep`.
```

mscdex commented on benchmark/README.md:
```diff
@@ -1,147 +1,276 @@
-var common = require('../common.js'); // Load the test runner
+Because the scatter plot can only show two variables (in this case _chunk_ and
+_encoding_) the rest is aggregated. Sometimes aggregating is a problem, this
+can be solved by filtering. This can be done during benchmarking by using
```
I think this same sentence can be reworded similarly as suggested earlier.

mscdex commented:
/cc @nodejs/collaborators

ChALkeR commented:
@mscdex What's the semver status of this? Major?

mscdex commented:
@ChALkeR I don't know how benchmarks are covered when it comes to that kind of thing. I would guess they are treated like tests or docs since they are not a part of the runtime?

mcollina commented:
I'll go for major, it makes things easier and less complicated.

One thing that is not clear from the document is how the statistical significance is achieved.

AndreasMadsen commented:
@mscdex Thanks for the suggestions, I will update the documentation tomorrow.

@mcollina It runs each the benchmark a given number of times (`--runs`) using the new and old node binary that is provided to `compare.js`. Using the R script it then:

> makes an independent/unpaired 2-group t-test, with the null hypothesis that the performance is the same for both versions. The significant field will show a star if the p-value is less than 0.05.

I think the [compare documentation](https://github.com/AndreasMadsen/node/blob/benchmark-refactor/benchmark/README.md#comparing-node-versions) is fairly clear on this. But do tell me how I can improve it.

mcollina commented:
I do not know what a independent/unpaired 2-group t-test is, I think providing a link might be helpful. A link is probably just fine as well.
I will explode this into its own headline, just because it will be easier to browse.

Also, add this into the PR description.

trevnorris commented on benchmark/_cli.js:
```diff
@@ -0,0 +1,93 @@
+
+  if (process.argv.length < 3) {
+    this.abort(usage);
+    return;
```
`.abort()` runs `process.exit(1)`. no need for the `return`. unless you're using it as a visual queue or something.

trevnorris commented on benchmark/_cli.js:
```diff
@@ -0,0 +1,93 @@
+      currentOptional = null;
+    }
+    // Required arguments
+    else if (!detectOptionals) {
```
are we running lint on these files?

AndreasMadsen replied:
`make lint` passes so I guess not.

AndreasMadsen replied:
Correction `./benchmark` is linted, this is apparently valid. https://github.com/nodejs/node/blob/master/Makefile#L665

trevnorris replied:
/cc @Trott @silverwind Didn't think the next line `else` was allowed in lint. It's not done anywhere else in core, and only 14 places in tests. Has been a rule, to the best of my knowledge, that has been at least informally enforced. There an option for this in the linter?

Trott replied:
I think we can enforce this with the `brace-style` rule. We're not currently using it. I'll see about getting a PR together.

Trott replied:
And...I take it back. I think it's do-able. I'll stop cluttering this issue now.

trevnorris commented:
did partial review. will finish up later tonight.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
-  }
+// parse configuarion arguments
+const configArgs = {};
+for (const arg of process.argv.slice(2)) {
```
not that it matters in this case, but remember that it evaluates `.slice(2)` every time.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
+for (const arg of process.argv.slice(2)) {
+  const match = arg.split('=');
+  const isNumber = isFinite(match[1]) && match[1] !== '';
+  configArgs[match[0]] = isNumber ? parseFloat(match[1]) : match[1];
```
not sure how much we care about this, but doing the check so simply arguments like `--foo=42=bar` will pass. May consider switching it to `arg.split(/^([^=]+)=/)` or something similar.

AndreasMadsen replied:
You are absolutely right. I will keep the original match logic.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
 }
 
+// The port used by servers and wrk
+exports.PORT = process.env.PORT || 12346;
```
is this a new `env`? if so then i'd suggest we switch it to `NODE_PORT`. just `PORT` has greater likelihood of conflicts with existing env variables.

AndreasMadsen replied:
I agree, but current version uses this too: https://github.com/nodejs/node/blob/master/benchmark/common.js#L16

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
-  var out = '';
-  var child = child_process.spawn('wrk', args);
+  const childStart = process.hrtime();
+  const child = child_process.spawn('wrk', args);
```
I question this. If the http tests don't run long enough to offset the time it took to startup the process then you have a problem. Couldn't you start timing on the first message received from the child process' `stdout`?

AndreasMadsen replied:
The benchmark don't directly use `process.hrtime()` in the http case, but instead the op/sec reported by `wrk` in the stdout. The `childStart` stuff is just for reporting how much time the benchmark actually use. This is only used as a side metic, which I found useful for finding benchmarks that uses to many iterations.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
 
+  // Collect stdout
+  let stdout = '';
   child.stdout.setEncoding('utf8');
```
not a lot of data is going through this so don't think it matters, but `setEncoding()` adds pretty hefty overhead to data events. i don't see why we can't just do `chunk.toString()` instead.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
-  child.on('close', function(code) {
-    if (cb)
-      cb(code);
+  child.once('close', function(code) {
```
do you know the time difference between the final `data` event and the `close` event? doubt it's much, but want to remove any static in the test.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
+      console.error('wrk produced strange output:');
+      console.error(stdout);
       process.exit(1);
+      return;
```
again, why with the `return` after a `process.exit()`?

AndreasMadsen replied:
good question :)

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
-    self.report(+qps);
+
+    // Report rate
+    self.report(parseFloat(match[1]), elapsed);
```
not that we'd use it, but this'll fail if the string is in hex/binary. instead simply using `+match[1]` will properly coerce those values.

trevnorris commented on benchmark/common.js:
```diff
@@ -1,205 +1,207 @@
+
+  // If there is only one configuarion don't spawn a process, just run
+  // the benchmark function
+  if (this.queue.length === 1) return this.fn(this.queue[0]);
```
i feel like we should always spawn a process in order to keep measurements uniform as possible. just my opinion though.

AndreasMadsen replied:
Good idea

trevnorris commented on benchmark/fs/readfile.js:
```diff
@@ -25,7 +25,7 @@ function main(conf) {
   setTimeout(function() {
     bench.end(reads);
-    try { fs.unlinkSync(filename); } catch (e) {}
+    process.exit(0);
```
why no longer try to unlink the file?

AndreasMadsen replied:
ups, this was from some debugging. But the line after `bench.end(reads);` never executed before, because it called `process.exit(0)`.

I will update such it removes the file.

trevnorris commented on benchmark/scatter.js:
```diff
@@ -0,0 +1,63 @@
+    // print csv header
+    if (printHeader) {
+      const confHeader = Object.keys(data.conf).join('", "');
+      console.log(`"filename", "${confHeader}", "rate", "time"`);
```
`"${confHeader}"`? did you mean `` `${confHeader}` ``

AndreasMadsen replied:
No but I understand the confusion. Is this easier to understand?

```js
      const confHeader = '"' + Object.keys(data.conf).join('", "') + '"';
      console.log(`"filename", ${confHeader}, "rate", "time"`);
```

trevnorris replied:
Sorry, still confused. What's in the PR and what you have in your example look like they'll have different output to me. Mind showing me an example of what would be logged?

AndreasMadsen replied:
Here is an example from running `./node benchmark/scatter.js benchmark/arrays/var-int.js`, the confusing part is inserting `"n", "type"` into the header.

```
"filename", "n", "type", "rate", "time"
"arrays/var-int.js", 25, Array, 73.73121801182165, 0.339069402
"arrays/var-int.js", 25, Buffer, 93.95908515556611, 0.266073259
"arrays/var-int.js", 25, Int8Array, 86.30949711522933, 0.289655262
"arrays/var-int.js", 25, Uint8Array, 95.8215561613897, 0.260901628
"arrays/var-int.js", 25, Int16Array, 86.45278377064632, 0.289175188
"arrays/var-int.js", 25, Uint16Array, 89.49914833863428, 0.279332267
"arrays/var-int.js", 25, Int32Array, 90.51250166094967, 0.276204939
```

trevnorris replied:
Okay, what I thought. What confuses me is that the PR uses `"${confHeader}"` where I would have expected it to look like your alternative suggestion using `` `${confHeader}` ``. Though I'm not sure why you have text in quotes and are using comma delimited format.

AndreasMadsen replied:
The output is a `.csv` file, which is easy to read from `R`. CSV usually doesn't require quotes, but if a string contains `,` quotes must be used.

AndreasMadsen replied:
There is now a `csvEncodeValue` function which should make the intention clear.

trevnorris commented:
finished initial review. thanks for all the work here. good stuff.

AndreasMadsen commented:
rebased and updated. But I still need to check that all the benchmarks works, which takes a very long time.

AndreasMadsen commented:
Finished the last bit of corrections

* The csv was misformatted after adding `JSON.stringify`
* if no benchmarks was found, it now prints a meaningful error
* changed `--var` to `--set`, I think it is more intuitive

trevnorris commented:
Just have one question, and would like to confirm that the JS files are linted with `make jslint` (did they before?). Other than that the change LGTM.

AndreasMadsen commented:
Here is the output from `make jslint`:

```
make jslint
./node tools/jslint.js -J benchmark lib src test tools/doc \
  tools/eslint-rules tools/jslint.js
[00:07|%100|+  1606|-     0]: Done
```

PS: I think we will have to solve https://github.com/nodejs/node/pull/7311 before we can land this.

mscdex commented on benchmark/scatter.js:
```diff
@@ -0,0 +1,63 @@
+    // print data row
+    const confData = Object.keys(data.conf)
+      .map((key) => data.conf[key])
+      .join(', ');
```
This may need to be improved a bit as far as escaping goes, at least in case there are quotes in any of the inputs. I'm not sure how R parses CSV, but it may also have issues with other characters, including control characters (especially CR and LF which may be used in some benchmarks -- namely http).

AndreasMadsen replied:
good catch. If the csv follows the RFC standard R shouldn't have any trouble. This means the string should just be surrounded with `"` and `"` should be escaped as `""`.

AndreasMadsen commented:
I've updated the code such it uses `child.send` to communicate the configuration object, as I sugested in #7311. I think this is the ideal solution as it does exactly what the user intents.

The `--set` option still have to guess if the value is a number or a string. I just set it to use the old `parseFloat` behaviour, as that has worked for a long time and I believe actual numbers are more common than numeric strings. For example I often use `--set n=10` to test the benchmark quickly. This behaviour is also much more transparent now, as the output surrounds strings with `"`. Thus the user can see how the value was parsed.

AndreasMadsen commented:
@trevnorris I've updated the code as you sugested in #7311.

The process arguments are now parsed by inferring the type from the options object and there is a validation that prevents mixing types in the same parameter.

I've reverted the `child.send()` changes, such it always communicates the configuration using process arguments. I see no reason to use both communication methods, now that the process arguments consistently works.

AndreasMadsen commented:
> The process arguments are now parsed by inferring the type from the options object and there is a validation that prevents mixing types in the same parameter.

@mscdex is this behaviour reasonable?

mscdex commented:
@AndreasMadsen Well, I think that would definitely be a step up from the current behavior.

Fishrock123 commented on benchmark/README.md:
```diff
@@ -1,147 +1,280 @@
+Each line represents a single benchmark with parameters specified as
+`${variable}=${value}`. Each configuration combination is executed in a separate
+process. This ensures that benchmark results aren't affected by the execution
+order due to v8 optimizations. **The last number is the rate of operations
```
"rate of operations"?

AndreasMadsen replied:
This is the term used in the current README as well: https://github.com/nodejs/node/blob/master/benchmark/README.md#run-all-tests-of-a-given-type

> The last number is the rate of operations. Higher is better.

An operation is an iteration or similar, the rate is `operations / sec`. How can I clarify this?

Fishrock123 commented on benchmark/README.md:
````diff
@@ -1,147 +1,280 @@
+you performed in the benchmark.
 
-var SlowBuffer = require('buffer').SlowBuffer;
+```javascript
````
can be ```js

Fishrock123 commented:
Seems ok to me... but do these now require `R`? Should we link to how to install that?

addaleax commented:
We’re having R scripts in `benchmark/` right now, too… not that a note in the README wouldn’t be helpful anyway. :)

Fishrock123 commented:
@addaleax the benchmark comparison currently does not _require_ R though, it seems to with this although I could be mistaken...

addaleax commented on benchmark/compare.js:
```diff
@@ -1,181 +1,85 @@
+for (const filename of benchmarks) {
+  for (const binary of binaries) {
+    for (let iter = 0; iter < runs; iter++) {
+      queue.push({ binary, filename, iter });
```
I’ve been trying this PR out for #7425, and I’ve seen results marked as significant (`p` of the order of 10^-10) on benchmarks that weren’t even remotely touched (with only very slight changes in measured performance, < 1 %). I suspect background noise to be the reason for that, any I’m wondering if it would help to move the `iter` loop outwards here?

AndreasMadsen replied:
That sounds strange do you remember which benchmark?

> I’m wondering if it would help to move the iter loop outwards here?

This would only make a difference if something external is affecting the benchmark, in which case you are in trouble anyway.

addaleax replied:
> That sounds strange do you remember which benchmark?

`http/check_is_http_token` stood out in particular.

> This would only make a difference if something external is affecting the benchmark, in which case you are in trouble anyway.

Well, yeah, that’s what I’m thinking, too. But I’m afraid perfectly isolated benchmarking is something more theoretically achievable than practically (I’ve picked the machine with the least system load that I have access to).

AndreasMadsen replied:
You are absolutely correct, I compared the same node build with itself and systematically watched a movie while it was running the "old" build and got an "improvement" and left it free when running the "new" build. This showed statistical significance, beyond what can be expected from a 95% confidence interval.

I then did the same without any load and got 2/32 statistical significant tests, as we would expected. The p-value was also much closer to 0.05.

Doing the iteration as the outer loop is a good idea. Alternatively it should be done randomly. I think both solutions will be fine.

addaleax commented on benchmark/compare.R:
```diff
@@ -0,0 +1,70 @@
+  old_mu = mean(subset(subdat, binary == "old")$rate);
+  improvement = sprintf("%.2f %%", ((new_mu - old_mu) / old_mu * 100));
+
+  # Add userfrindly stars to the table. There should be at least one star
```
typo: `userfrindly`

AndreasMadsen commented:
> the benchmark comparison currently does not require R though, 

`R` is not a new requirement, but you are correct that it wasn't required for `compare.js` before.

> Should we link to how to install that?

Most package managers will provide this. The R webpage doesn't have good install instructions, it just provides a long list of mirrors. https://www.r-project.org/ - How should we document installation?

trevnorris commented:
Thanks for addressing all my comments. I dig the changes.

While it can't/shouldn't be addressed here, want devs to understand performance benchmarks are still more complicated than this. For example, how would the fs operations perform if a child process was spawned that performed random amounts of disk activity while the benchmark ran. Oh well. Maybe just need to add more tests. This is a great step in the right direction. Thanks for all your work here.

trevnorris commented:
@AndreasMadsen Anything preventing this from landing?

Fishrock123 commented:
Rubber-stamp LGTM *if* i am holding this up. I only assigned myself to keep it on my radar. I like the changes in premise but do not have time to review them in detail.

(I may be able to make time to try them if desired though.)

AndreasMadsen commented:
@trevnorris @Fishrock123 I was on vacation, before that I was waiting for #7630 to land.

I have fixed the lint and merge conflict issues. As this is a rather large change, I'm won't merge before Thusday Morning (UTC+0), such people have time to come with last objections. (72 hours as suggested in [onboarding.md](https://github.com/nodejs/node/blob/master/doc/onboarding.md))

I plan to tag @mscdex, @trevnorris, @Fishrock123, @addaleax as reviewers on this PR, if there are no objections.

addaleax commented on benchmark/README.md:
```diff
@@ -1,147 +1,283 @@
-Most of the http benchmarks require [`wrk`][wrk] and [`ab`][ab] (ApacheBench) being installed.
-These may be available through your preferred package manager.
+Most of the http benchmarks require [`wrk`][wrk] to be installed. It may be
+available through your preferred package manager. If not `wrk` can be built
```
nit: comma after `not`

addaleax commented on benchmark/README.md:
```diff
@@ -1,147 +1,283 @@
-### Run tests with options
+In the output _improvement_ is the relative improvement of the new version,
+hopefully this is positive. _significant_ tells if there are enough
+statistical evidence to validate the _improvement_. If there are enough evidence
```
nit: Both occurrences of `evidence` should come with singular verbs.

addaleax commented on benchmark/README.md:
```diff
@@ -1,147 +1,283 @@
+_improvement_.** Sometimes this is fine, for example if you are expecting there
+to be no improvements, then there shouldn't be any stars.
+
+_For the statistically minded the R scripts makes an [independent/unpaired
```
nit: `For the statistically minded, the R script performs an`?

AndreasMadsen replied:
Can you elaborate on what is wrong, it says:

> For the statistically minded the R scripts makes an independent/unpaired 2-group t-test, with the null hypothesis that the performance is the same for both versions. The significant field will show a star if the p-value is less than 0.05.

addaleax replied:
So, I’m not a native speaker, but I would say there should be a comma after `,`; otherwise, the sentence sounds like the R script does what it does *for specific people*. :)

The `scripts` -> `script` change suggestion is only to make the noun match the singular verb, and `makes` -> `performs` is just a suggestion based on how I think most people use the language around statistical tests.

AndreasMadsen commented:
I've updated the CLI API such the `--` options comes first, hopefully this will be less confusing.

```
usage: ./node compare.js [options] [--] <category> ...
  Run each benchmark in the <category> directory many times using two diffrent
  node versions. More than one <category> directory can be specified.
  The output is formatted as csv, which can be processed using for
  example 'compare.R'.

  --new    ./new-node-binary  new node binary (required)
  --old    ./old-node-binary  old node binary (required)
  --runs   30                 number of samples
  --filter pattern            string to filter benchmark scripts
  --set    variable=value     set benchmark variable (can be repeated)
```

```
usage: ./node scatter.js [options] [--] <filename>
  Run the benchmark script <filename> many times and output the rate (ops/s)
  together with the benchmark variables as a csv.

  --runs 30              number of samples
  --set  variable=value  set benchmark variable (can be repeated)
```

```
usage: ./node run.js [options] [--] <category> ...
  Run each benchmark in the <category> directory a single time, more than one
  <categoty> directory can be specified.

  --filter pattern          string to filter benchmark scripts
  --set    variable=value   set benchmark variable (can be repeated)
```

AndreasMadsen commented:
I had a short conversation with @addaleax about how we should handle type 1 error, we agreed that a note in the documentation should be sufficient. I've added the following

> **a word of caution:** Statistics is not a foolproof tool. If a benchmark shows a statistical significant difference, there is a 5% risk that this difference doesn't actually exists. For a single benchmark this is not an issue. But when considering 20 benchmarks it's normal that one of them will show significance, when it shouldn't. A possible solution it to instead consider at least two stars (`**`) as the threshold, in that case the risk is 1%. If three stars (`***`) is considered the risk is 0.1%. However this may require more runs to obtain (can be set with --runs).

Trott commented on benchmark/README.md:
```diff
@@ -1,147 +1,292 @@
+_improvement_.** Sometimes this is fine, for example if you are expecting there
+to be no improvements, then there shouldn't be any stars.
+
+**a word of caution:** Statistics is not a foolproof tool. If a benchmark shows
```
Nit: `a word of caution` -> `A word of caution:`

Trott commented on benchmark/README.md:
```diff
@@ -1,147 +1,292 @@
+a statistical significant difference, there is a 5% risk that this
+difference doesn't actually exists. For a single benchmark this is not an
+issue. But when considering 20 benchmarks it's normal that one of them
+will show significance, when it shouldn't. A possible solution it to instead
```
Typo: `possible solution it to instead` -> `possible solution is to instead`

AndreasMadsen commented:
Thanks for the review. Landed in ee2843b4ea54b04eca30e00ec1ab39206449bd52 edbed3f3fd8b0010f3f474934ad880fa2898265a 0f9bfaa7c51bb5f4910ddf068ed89b1bdf35e7d5 f3463cf3061931b5c94ba9c753c1d75ee4d2b712 1f64ceba89a074f9e23196d019d56f00cdd4577a 01fbf656a3874d189cadeced08266a26ea526491 de9b44c0889d2264436277848762f1ebf868aa57 6e745d7a7586b12b894537192726bf2b999a456d 693e7be399e4c0964b5bbceaee6e8326c7c02a42

addaleax commented:
Uh, you might want to back these commits out of master for now, the linter complains about `benchmark/_cli.js`

AndreasMadsen commented:
As in force push?

addaleax commented:
@AndreasMadsen I did that for now. Could you fix that, and maybe do a CI or linter run before re-landing? ;)

AndreasMadsen commented:
Thanks for the quick eye. I have force pushed and updated the PR. I wish I know how it happened.

CI: https://ci.nodejs.org/job/node-test-pull-request/3422/

addaleax commented:
Well, yeah, I’ve had the, ahem, pleasure of breaking master by not having run CI again before landing myself in the recent past. :)

Anyway, CI looked good before it went all 502 (FreeBSD failure is unrelated and only the Windows tests were remaining), I’d say you can land this. Thanks!

AndreasMadsen commented:
Landed in: ee2843b4ea54b04eca30e00ec1ab39206449bd52 edbed3f3fd8b0010f3f474934ad880fa2898265a 0f9bfaa7c51bb5f4910ddf068ed89b1bdf35e7d5 f99471b2ae147fbb072223be62e9100862379dc8 8bb59fdb127b060d38430c0428d7783931ebec6e 855009af7f46c3fb7c5653713a0930e0b9c33f45 0c0f34e2fe0e73fdbad3c768a7497032f91f5b20 6edef1deb95f915da2101d1af557a4b72fbf16af d525e6c92aa4a5806a12cfc9206e85724cafa9bc

addaleax commented:
Labelled this semver-major because that’s what has been suggested above, and #7890 shows that people obviously were using APIs of the old benchmarking scripts.

AndreasMadsen commented:
Sounds good. This is obviously not backward compatible and it is quite easy to use the new tools on an old node version.

_Also I don't really want to backport this ;)_
